### PR TITLE
Caliper: some nonconformant function type conversion needs -fpermissive to compile (#8549)

### LINKED ISSUE #8549: Fail to build caliper on Quartz
Summary
Spack fails to build caliper on Quartz.

### Expected Result

Build succeeds.

### Actual Result
Sample error message: 

10 errors found in build log:
     342    make -f ext/gotcha/gotcha-download/gotcha-build/src/example/minimal/CMakeFiles/wrap_me.dir/build.make ext/gotcha/gotcha-download/gotcha-build/src/example/minima
            l/CMakeFiles/wrap_me.dir/build
     343    make[2]: Leaving directory `/tmp/wang102/spack-stage/spack-stage-XxzksK/Caliper/spack-build'
     344    make -f ext/gotcha/gotcha-download/gotcha-build/src/example/autotee/CMakeFiles/autotee.dir/build.make ext/gotcha/gotcha-download/gotcha-build/src/example/autote
            e/CMakeFiles/autotee.dir/build
     345    make[2]: Entering directory `/tmp/wang102/spack-stage/spack-stage-XxzksK/Caliper/spack-build'
     346    make[2]: Entering directory `/tmp/wang102/spack-stage/spack-stage-XxzksK/Caliper/spack-build'
     347    /usr/workspace/wsa/wang102/spack/var/spack/stage/caliper-1.6.0-c7bkii4x5qlvcuanjxxqop36ukvxdnop/Caliper/src/services/pthread/PthreadService.cpp: In function 'vo
            id {anonymous}::pthreadservice_initialize(cali::Caliper*)':
  >> 348    /usr/workspace/wsa/wang102/spack/var/spack/stage/caliper-1.6.0-c7bkii4x5qlvcuanjxxqop36ukvxdnop/Caliper/src/services/pthread/PthreadService.cpp:94:5: error: inv
            alid conversion from 'int (**)(pthread_t*, const pthread_attr_t*, void* (*)(void*), void*) {aka int (**)(long unsigned int*, const pthread_attr_t*, void* (*)(vo
            id*), void*)}' to 'void**' [-fpermissive]
     349         };

### Steps to reproduce the issue

spack install caliper

### Information on your system
Machine: Quartz
Spack branch: develop cea5d04 Fix GDAL Python bindings and add import tests (#8482)

### PR DESCRIPTION
**Pull request by twang15**
Fixes #8549

### CODE REVIEW

File: var/spack/repos/builtin/packages/caliper/package.py [hunk 1]
```diff
@@ -96,4 +96,8 @@ def cmake_args(self):
             args.append('-DMPI_C_COMPILER=%s' % spec['mpi'].mpicc)
             args.append('-DMPI_CXX_COMPILER=%s' % spec['mpi'].mpicxx)
 
+        # allow some nonconforming code to compile
+        compile_flags = "-O3 -fpermissive"
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/caliper/package.py]: These flags will only work for GCC, not other compilers.

File: var/spack/repos/builtin/packages/caliper/package.py [hunk 2]
```diff
@@ -96,4 +96,8 @@ def cmake_args(self):
...
+        args.append('-DCMAKE_CXX_FLAGS=%s' % compile_flags)
```
Comment (adamjstewart) [inline var/spack/repos/builtin/packages/caliper/package.py]: See #8543 for an example of a more robust way to handle this.

**Comment by daboehme** [comment]
@twang15 which compiler gave you that error and in which file?

**Comment by daboehme** [comment]
Oh, you're building 1.6.0. Unfortunately we made a bit of a mistake there with our GOTCHA integration. Caliper 1.6.0 is downloading a now-incompatible GOTCHA version, which will likely produce this compile error. The `-fpermissive` switch may make it compile, but it won't work. This will be fixed in 1.7.0. I'll upload an updated spack package soon. Meanwhile you can either install the current master (`spack install caliper@master`) or replace the 1.6.0 in the spack package with 1.7.0.

**Comment by adamjstewart** [comment]
> Caliper 1.6.0 is downloading a now-incompatible GOTCHA version

Do you mean that the Spack `caliper` package downloads and installs its own dependencies during installation? Is there any way to prevent that? It would be good to have a separate Spack package for `gotcha` that Spack can manage as a dependency. Many of our users are on air-gapped networks, so they won't have internet access during installation.

**Comment by daboehme** [comment]
Yes, in 1.6 Caliper did download gotcha during its installation process. In 1.7 there's an option to use a pre-installed gotcha. I can change the Caliper spack package so that it uses gotcha as a spack dependency.

**Comment by adamjstewart** [comment]
Superseded by #8572.

**Pull request closed without merging**